modules/standard/Lists.chpl

```diff
@@ -0,0 +1,363 @@
+    - proc push_front(vals: [])
+    - proc pop_front()
+    - proc insert(pos: this.idxType, in val: this.eltType)
+    - proc insert(pos: this.idxType, vals: [])
+    - proc remove(pos: this.idxType)
+    - proc remove(pos: this.idxType, count: this.idxType)
+    - proc remove(pos: range(this.idxType, stridable = false))
+    - proc clear()
+
+  All of the methods given above change an array's domain (exactly what we
+  are trying to avoid, and the impetus for the List type).
+
+*/
+module Lists {
+
+  // Building blocks used to store "chunks" of List elements.
+  pragma "no doc"
+  class ListBlock {
+    type _etype;
+    var data: _ddata(_etype) = nil;
+  }
+
+  /*
+    A container intended to replace the use of arrays to perform "vector-like"
+    operations.
+
+    Is not currently parallel safe.
+
+    The following operations may trigger a move:
+      - insert
+      - remove
+      - reverse
+      - sort
+      - pop
+
+    A move will invalidate any references to List elements held by tasks, as
+    well as iterators to List elements produced by tasks before the move
+    occurred.
+  */
+  class List {
+
+    pragma "no doc"
+    var _size = 0;
+    pragma "no doc"
+    var _etype;
+
+    //
+    // Michael suggested that we move from a LL of blocks to an array, which
+    // would give us constant speed indexing at the cost of more complicated
+    // indexing logic.
+    //
+    // For now, I can start with Vass's logic and then switch over to a O(1)
+    // indexing scheme when everything else is working.
+    //
+    pragma "no doc"
+    var _head, _tail: unmanaged ListBlock(_etype) = nil;
+
+    /*
+      Initializes a new List containing elements of the given type.
+
+      :arg etype: The type of the elements of this List.
+    */
+    proc init(type etype) {
+      _etype = etype;
+    }
+
+    /*
+      Initializes a new List as a copy of another List object.
+
+      :arg other: The List object to initialize from.
+    */
+    proc init=(other: List) {
+      this._etype = other._etype;
+    }
+  
+    proc deinit() {}
+
+    /*
+      Add an element to the end of this List.
+
+      .. note::
+
+        Appending an element to a List will never invalidate references to
+        elements in the List.
+
+      :arg x: An element to append.
+    */
+    proc append(x: this._etype) {}
+
+    /*
+      Extend this List by appending all the elements from another List.
+
+      :arg other: A List, the elements of which are appended to this List.
+    */
+    proc extend(other: List(this._etype)) {}
+
+    /*
+      Extend this List by appending all the elements from an array.
+
+      :arg other: An array, the elements of which are appended to this List.
+    */
+    proc extend(other: this._etype[?d]) {}
+
+    /*
+      Insert an element at a given position in this List, shifting all elements
+      following that index one to the right. Note that `a.insert(a.size, x)`
+      is equivalent to `a.append(x)`.
```

Comment:
Append or extend?

Reply:
Would be good to link to the functions you are mentioning, both here and later

Reply:
Good point!